drivers/tee/optee/optee.c

```diff
@@ -0,0 +1,500 @@
+{
+	int i;
+	const struct tee_param *tp = param;
+	struct optee_msg_param *mtp = msg_param;
+
+	if (!param || !msg_param) {
+		return -EINVAL;
+	}
+
+	for (i = 0; i < num_param; i++, tp++, mtp++) {
+		if (!tp || !mtp) {
+			LOG_ERR("Wrong param on %d iteration", i);
+			return -EINVAL;
+		}
+
+		switch (tp->attr) {
+		case TEE_PARAM_ATTR_TYPE_NONE:
+			mtp->attr = OPTEE_MSG_ATTR_TYPE_NONE;
+			memset(&mtp->u, 0, sizeof(mtp->u));
+			break;
+		case TEE_PARAM_ATTR_TYPE_VALUE_INPUT:
+		case TEE_PARAM_ATTR_TYPE_VALUE_OUTPUT:
+		case TEE_PARAM_ATTR_TYPE_VALUE_INOUT:
+			mtp->attr = OPTEE_MSG_ATTR_TYPE_VALUE_INPUT + tp->attr -
+				TEE_PARAM_ATTR_TYPE_VALUE_INPUT;
+			mtp->u.value.a = tp->a;
+			mtp->u.value.b = tp->b;
+			mtp->u.value.c = tp->c;
+			break;
+		case TEE_PARAM_ATTR_TYPE_MEMREF_INPUT:
+		case TEE_PARAM_ATTR_TYPE_MEMREF_OUTPUT:
+		case TEE_PARAM_ATTR_TYPE_MEMREF_INOUT:
+			mtp->attr = OPTEE_MSG_ATTR_TYPE_RMEM_INPUT + tp->attr -
+				TEE_PARAM_ATTR_TYPE_MEMREF_INPUT;
+			mtp->u.rmem.shm_ref = tp->c;
+			mtp->u.rmem.size = tp->b;
+			mtp->u.rmem.offs = tp->a;
+			break;
+		default:
+			return -EINVAL;
+		}
+	}
+
+	return 0;
+}
+
+static int msg_param_to_param(struct tee_param *param, unsigned int num_param,
+			      const struct optee_msg_param *msg_param)
+{
+	int i;
+	struct tee_param *tp = param;
+	const struct optee_msg_param *mtp = msg_param;
+
+	if (!param || !msg_param) {
+		return -EINVAL;
+	}
+
+	for (i = 0; i < num_param; i++, tp++, mtp++) {
+		uint32_t attr = mtp->attr & OPTEE_MSG_ATTR_TYPE_MASK;
+
+		if (!tp || !mtp) {
+			LOG_ERR("Wrong param on %d iteration", i);
+			return -EINVAL;
+		}
+
+		switch (attr) {
+		case OPTEE_MSG_ATTR_TYPE_NONE:
+			memset(tp, 0, sizeof(*tp));
+			tp->attr = TEE_PARAM_ATTR_TYPE_NONE;
+			break;
+		case OPTEE_MSG_ATTR_TYPE_VALUE_INPUT:
+		case OPTEE_MSG_ATTR_TYPE_VALUE_OUTPUT:
+		case OPTEE_MSG_ATTR_TYPE_VALUE_INOUT:
+			tp->attr = TEE_PARAM_ATTR_TYPE_VALUE_INOUT + attr -
+				OPTEE_MSG_ATTR_TYPE_VALUE_INPUT;
+			tp->a = mtp->u.value.a;
+			tp->b = mtp->u.value.b;
+			tp->c = mtp->u.value.c;
+			break;
+		case OPTEE_MSG_ATTR_TYPE_RMEM_INPUT:
+		case OPTEE_MSG_ATTR_TYPE_RMEM_OUTPUT:
+		case OPTEE_MSG_ATTR_TYPE_RMEM_INOUT:
+			tp->attr = TEE_PARAM_ATTR_TYPE_MEMREF_INPUT + attr -
+				OPTEE_MSG_ATTR_TYPE_RMEM_INPUT;
+			tp->b = mtp->u.rmem.size;
+
+			if (!mtp->u.rmem.shm_ref) {
+				tp->a = 0;
+				tp->c = 0;
+				break;
+			}
+
+			tp->a = mtp->u.rmem.offs;
+			tp->c = mtp->u.rmem.shm_ref;
+			break;
+		default:
+			return -EINVAL;
+		}
+	}
+
+	return 0;
+}
+
+static uint64_t regs_to_u64(uint32_t reg0, uint32_t reg1)
+{
+	return (uint64_t)(((uint64_t)reg0 << 32) | reg1);
+}
+
+static void u64_to_regs(uint64_t val, uint32_t *reg0, uint32_t *reg1)
+{
+	*reg0 = val >> 32;
+	*reg1 = val;
+}
+
+static void handle_rpc_call(const struct device *dev, struct optee_rpc_param *param)
+{
+	struct tee_shm *shm = NULL;
+
+	switch (OPTEE_SMC_RETURN_GET_RPC_FUNC(param->a0)) {
```

Comment:
I'm not sure if it's really needed. optee_call pass &param to the local variable. And it's impossible to get NULL here. I don't mind about checks but thought it's not necessary here. I think it's better to avoid unneeded checks.